tests/sentry/api/endpoints/test_project_details.py

```diff
@@ -752,7 +752,8 @@ def test_cap_secondary_grouping_expiry(self):
         expiry = response.data["secondaryGroupingExpiry"]
         assert (now + 3600 * 24 * 90) < expiry < (now + 3600 * 24 * 92)
 
-    def test_redacted_symbol_source_secrets(self):
+    @mock.patch("sentry.utils.audit.create_audit_entry")
```

Comment:
I'm botching something here I think, and I can't figure out what it is right now

Reply:
So python is finicky^Wfun.  The function is called from `sentry.api.base` via some indirections.  If you look at that module it imports the function directly: `from sentry.utils.audit import create_audit_entry`.  Now at **import time** the globals of `sentry.api.base` module get the entry for `create_audit_entry`.

So sometime later while executing your test the monkeypatch is changing the `sentry.utils.audit.__globals__['create_audit_entry'] = patched_version`.  This will no longer affect `sentry.api.base.__globals__['create_audit_entry'] because import time is long over and the actual function you are trying to test keeps using the unpatched version.  The solution is to patch the thing which is actually being executed during the test: `sentry.api.base.create_audit_entry`.

(as an aside, there's some old style-and-or-convention to prefer importing entire modules in python rather than things inside modules, but it doesn't seem to be followed in sentry much.  In that case the import in `sentry.api.base` would have been `from sentry.utils import audit` and the call `audit.create_audit_entry`, in that case monkeypatching `sentry.utils.audit.create_audit_library` would have affected the globals of the audit module as expected and the code does the function lookup in the globals of the module at runtime and would have found the monkeypatched version.  oh well)

In any case, I committed a change to fix this test.  I also did some renaming to make me less confused.  Lastly note that the test had to be changed to not check for the serialised config since it is now put in the audit log without serialising.  I think this is ok for behaviour so adapted the test rather than the code.

Reply:
thanks for fixing this and the in-depth explanation! is there a document that goes into detail how these rules work in python when it comes to imports (e.g. which module is the true owner of an imported or exported method), or is this particular `create_audit_entry` instance an example of sentry monkeypatching and doing something that's specific to sentry itself?

Reply:
No this isn't specific to sentry, not sure if there's a simple document.  it's kind of a side effect of how the entire python execution model is built and everything is done by an interpreter one thing after the other, how globals and locals are just dicts etc.